pkg/ottl/ottlfuncs/func_append.go

```diff
@@ -0,0 +1,134 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type AppendArguments[K any] struct {
+	Target ottl.GetSetter[K]
+	Value  ottl.Optional[ottl.Getter[K]]
+	Values ottl.Optional[[]ottl.Getter[K]]
+}
+
+func NewAppendFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("append", &AppendArguments[K]{}, createAppendFunction[K])
+}
+func createAppendFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*AppendArguments[K])
+	if !ok {
+		return nil, fmt.Errorf("AppendFactory args must be of type *Appendrguments[K]")
+	}
+
+	return appendTo(args.Target, args.Value, args.Values)
+}
+
+func appendTo[K any](target ottl.GetSetter[K], value ottl.Optional[ottl.Getter[K]], values ottl.Optional[[]ottl.Getter[K]]) (ottl.ExprFunc[K], error) {
+	if value.IsEmpty() && values.IsEmpty() {
+		return nil, fmt.Errorf("at least one of the optional arguments must be provided be provided")
+	}
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		t, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		// init res with target values
+		var res []any
+
+		if t != nil {
+			switch targetType := t.(type) {
```

Comment:
I know it's a bit tedious, but could you add exhaustive tests for the types here (and the default cases if possible)? I just want to test that we don't encounter any errors when working with these types, we don't need to combine any complex invocations for the tests. I just want to make sure we're thorough, plus this will help set the stage for adding types in the future without risking errors.